// react-relay/lib/RelayOperationSelector.js
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 * @format
 */

'use strict';

/**
 * @public
 *
 * Implementation of `RelayCore#createOperationSelector()` defined in
 * `RelayEnvironmentTypes` for the classic core.
 */
function createOperationSelector(operation, variables,
// unused param for compatibility with modern API
_modernOperation) {
  var concreteFragment = require('./QueryBuilder').getFragment(operation.node);
  !concreteFragment ? process.env.NODE_ENV !== 'production' ? require('fbjs/lib/invariant')(false, 'RelayOperationSelector: Expected a query, got %s `%s`.', operation.node.kind, operation.name) : require('fbjs/lib/invariant')(false) : void 0;

  var operationVariables = require('./RelayVariables').getOperationVariables(operation, variables);
  var fragment = {
    dataID: require('./RelayStoreConstants').ROOT_ID,
    node: concreteFragment,
    variables: operationVariables
  };

  return {
    fragment: fragment,
    node: operation,
    root: fragment,
    variables: operationVariables
  };
}

module.exports = {
  createOperationSelector: createOperationSelector
};
