// relay-runtime/lib/ConvertToExecuteFunction.js
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 * @format
 */

'use strict';

var _extends3 = _interopRequireDefault(require('babel-runtime/helpers/extends'));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function filterDataFromStream(observable) {
  return require('./RelayObservable').create(function (sink) {
    return observable.subscribe({
      next: function next(value) {
        if (value.kind === 'data') {
          sink.next(value);
        }
      },

      error: sink.error,
      complete: sink.complete
    });
  });
}

/**
 * Converts a FetchFunction into an ExecuteFunction for use by RelayNetwork.
 */
function convertFetch(fn) {
  return function fetch(request, variables, cacheConfig, uploadables) {
    var result = fn(request, variables, cacheConfig, uploadables);
    // Note: We allow FetchFunction to directly return Error to indicate
    // a failure to fetch. To avoid handling this special case throughout the
    // Relay codebase, it is explicitly handled here.
    if (result instanceof Error) {
      return filterDataFromStream(require('./RelayObservable').create(function (sink) {
        return sink.error(result);
      }));
    }
    var withEvents = require('./RelayObservable').from(result).map(function (value) {
      return convertToStreamPayload(request, variables, value);
    });
    return filterDataFromStream(withEvents);
  };
}

/**
 * Converts a SubscribeFunction into an ExecuteFunction for use by RelayNetwork.
 */
function convertSubscribe(fn) {
  return function subscribe(operation, variables, cacheConfig) {
    var withEvents = require('./RelayObservable').fromLegacy(function (observer) {
      return fn(operation, variables, cacheConfig, observer);
    }).map(function (value) {
      return convertToStreamPayload(operation, variables, value);
    });
    return filterDataFromStream(withEvents);
  };
}

function convertSubscribeWithEvents(fn) {
  return function subscribe(operation, variables, cacheConfig) {
    return require('./RelayObservable').fromLegacy(function (observer) {
      return fn(operation, variables, cacheConfig, observer);
    }).map(function (value) {
      return convertToStreamPayload(operation, variables, value);
    });
  };
}

/**
 * Given a value which might be a plain GraphQLResponse, coerce to always return
 * an ExecutePayload. A GraphQLResponse may be returned directly from older or
 * simpler Relay Network implementations.
 */
function convertToStreamPayload(request, variables, value) {
  if (!value.data && !value.errors && value.response) {
    if (!value.operation) {
      process.env.NODE_ENV !== 'production' ? require('fbjs/lib/warning')(false, 'ConvertToExecuteFunction: execute payload contains response but ' + 'is missing operation.') : void 0;
      return createExecutePayload(request, variables, value.response);
    }
    // assume value is data if value.response defined, but no kind field given
    return (0, _extends3['default'])({}, value, { kind: 'data' });
  } else if (!value.data && !value.errors && value.kind === 'event') {
    return value;
  } else {
    // in this case we have a GraphQLResponse
    return createExecutePayload(request, variables, value);
  }
}

function createExecutePayload(request, variables, response) {
  if (request.kind === require('./RelayConcreteNode').BATCH_REQUEST) {
    throw new Error('ConvertToExecuteFunction: Batch request must return ExecutePayload.');
  }
  return { kind: 'data', operation: request.operation, variables: variables, response: response };
}

module.exports = {
  convertFetch: convertFetch,
  convertSubscribe: convertSubscribe,
  convertSubscribeWithEvents: convertSubscribeWithEvents
};
